packages/compass-components/src/hooks/use-stacked-component.tsx

```diff
@@ -0,0 +1,93 @@
+import type { ComponentType, ForwardedRef } from 'react';
+import React, { createContext, useContext, useMemo } from 'react';
+
+type StackedComponentProviderProps = {
+  zIndex?: number;
+};
+
+const StackedComponentContext =
+  createContext<StackedComponentProviderProps | null>(null);
+
+export const StackedComponentProvider = ({
+  zIndex,
+  children,
+}: StackedComponentProviderProps & { children: React.ReactNode }) => {
+  const value = useMemo(() => ({ zIndex }), [zIndex]);
+  return (
+    <StackedComponentContext.Provider value={value}>
+      {children}
+    </StackedComponentContext.Provider>
+  );
+};
+
+export const useStackedComponent = () => {
+  return useContext(StackedComponentContext);
+};
+
+type StackedComponentProps<T extends boolean> = T extends true
+  ? { popoverZIndex?: number }
+  : T extends false
+  ? { style?: React.CSSProperties }
+  : Record<string, never>;
+
+// TODO(LG-4109): This should be eventually supported by the LG design system
+const withBaseStyles = function <UsePopover extends boolean, ComponentProps>(
+  WrappedComponent: ComponentType<ComponentProps>,
+  usePopoverZIndex: UsePopover
+): ComponentType<ComponentProps> {
+  const ComponentWithStackedStyles = (
+    props: ComponentProps,
+    ref: ForwardedRef<ComponentType<ComponentProps>>
```

Comment:
To create a forward ref component you actually need to define the component with [forwardRef](https://react.dev/reference/react/forwardRef)